### PR TITLE
add qe page

### PR DESCRIPTION
**Pull request by terakilobyte**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-22414>, <https://jira.mongodb.org/browse/DOCSP-22131>
Staging - <https://docs-mongodbcom-staging.corp.mongodb.com/node/docsworker-xlarge/DOCSP-22414/fundamentals/encrypt-fields/qe/>

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?

### CODE REVIEW

File: source/fundamentals/qe.txt [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+For in-depth information on how {+qe+} works, see the
+:ref:`{+qe+} reference <qe-features>` documentation.
+
+Installation
+------------
+
+To get started with {+qe+} in your client application, you need:
+
+- the MongoDB Node.js driver
+- `mongodb-client-encryption <https://www.npmjs.com/package/mongodb-client-encryption>`__ version ``2.2.0-alpha.1`` or later
```
Comment (addaleax) [inline source/fundamentals/qe.txt:39]: It might be better to just say `2.2.0` here in the docs, we don’t really want people using alpha versions I imagine.

File: source/fundamentals/qe.txt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+:ref:`crypt_shared reference documentation <qe-crypt_shared>`.
+
+``mongocryptd``
+~~~~~~~~~~~~~~~
+
+``mongocryptd`` is launched automatically by the package, and is used for
+automatic encryption. ``mongocryptd`` communicates with
+``mongodb-client-encryption`` to automatically encrypt the information
+specified by a user-provided
+:manual:`JSON Schema </reference/security-client-side-automatic-json-schema/>`.
```
Comment (addaleax) [inline source/fundamentals/qe.txt:73]: This might be a bit confusing, since QE doesn’t do JSON Schema anymore (at least not explicitly)?

File: source/fundamentals/encrypt-fields/qe.txt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+.. code-block:: sh
+   :copyable: true
+
+   npm install mongodb-client-encryption@2.2.0 --save
+
+Shared Library
+~~~~~~~~~~~~~~
+
+The shared library is automatically used by the package, and is used
```
Comment (ccho-mongodb) [inline source/fundamentals/encrypt-fields/qe.txt:61]: Issue: unsure what "the package" is referring to. Is it the driver or mongodb-client-encryption?

File: source/includes/qe/make-data-key.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    {
+      unique: true,
+      partialFilterExpression: { keyAltNames: { $exists: true } },
+    }
+  );
+  const clientEnc = new ClientEncryption(keyVaultClient, {
+    keyVaultNamespace: keyVaultNamespace,
+    kmsProviders: kmsProviders,
+  });
+  const dek1 = await clientEnc.createDataKey(provider, {
```
Comment (ccho-mongodb) [inline source/includes/qe/make-data-key.js:43]: For most programming languages, the style guides recommend staying away from numbers in variable names.

File: source/includes/qe/make-data-key.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  const encClient = new MongoClient(uri, {
+    autoEncryption: {
+      keyVaultNamespace,
+      kmsProviders,
+      extraOptions,
+      encryptedFieldsMap,
+    },
+  });
+  await encClient.connect();
+  const newEncDB = encClient.db(secretDB);
```
Comment (ccho-mongodb) [inline source/includes/qe/make-data-key.js:70]: Issue: the variable names in this file could be more meaningful. Currently, they use some mixed, non-idiomatic naming conventions.

Suggestion: "DB" -> "Db" for camelcase consistency, remove "new" from the variable name, "encrypted" is better than "enc", etc.

File: source/includes/qe/insert-encrypted-document.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
+const { MongoClient } = require("mongodb");
+
+const eDB = "encryption";
+const eKV = "__keyVault";
+const secretDB = "medicalRecords";
+const secretCollection = "patients";
+const keyVaultNamespace = `${eDB}.${eKV}`;
+
+const fs = require("fs");
+const provider = "local";
```
Comment (ccho-mongodb) [inline source/includes/qe/insert-encrypted-document.js:10]: Issue: unused variable "provider"

**Review by ccho-mongodb — CHANGES_REQUESTED** [review]
Most comments apply to both of the guides, please cross-reference when addressing.

In general, I would want a reader new to CSFLE and QE to understand the following:
1) What do they do (ELI5)? What are the various pieces of info I need to do to encrypt a field?
2) What's the difference between CSFLE and QE?
3) What am I achieving in the examples? What's the difference between normal client and an encrypted one?

File: snooty.toml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 name = "node"
 title = "Node.js"
 intersphinx = [
-    "https://www.mongodb.com/docs/manual/objects.inv",
+    "https://www.mongodb.com/docs/upcoming/objects.inv",
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply snooty.toml:4]: I don't think there needs to be a ticket. We can use both `upcoming` and `manual` `.inv` files. New refs defined in `upcoming` will get picked up, and any existing refs will overwrite.

File: source/includes/auto-encryption-options.rst [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+   * - **keyVaultClient**
+     - **Type:** ``MongoClient``
+
+       **Description**
+       A ``MongoClient`` configured to connect to
+       the MongoDB instance hosting your {+key-vault-long+}. For more
+
+       If you omit the ``keyVaultClient`` option, the driver uses the
+       MongoDB instance specified to your ``MongoClient`` containing the
+       ``AutoEncryptionOpts`` configuration as the host of your key vault
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:20]: Suggestion: Since you have a source constant key-vault-long, use it for all instances or else it may generate maintenance issues in the future.

Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:20]: I'm not sure I understand this perfectly, but maybe reordering the related info could help, e.g.:

"When you omit this option, the driver chooses the MongoDB instance in your configuration that ..."

File: source/includes/auto-encryption-options.rst [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+   :header-rows: 1
+   :widths: 30 70
+
+   * - Name
+     - Description
+
+   * - **keyVaultClient**
+     - **Type:** ``MongoClient``
+
+       **Description**
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:14]: Suggestion: Separate lines for Description (bolded) and the descriptive (plain) text.
\

File: source/includes/auto-encryption-options.rst [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+       If you omit the ``keyVaultClient`` option, the driver uses the
+       MongoDB instance specified to your ``MongoClient`` containing the
+       ``AutoEncryptionOpts`` configuration as the host of your key vault
+       collection.
+
+       To learn more about {+key-vault-long+}s, see :ref:`qe-reference-key-vault`.
+
+       **Default:** ``undefined``
+       **Accepted Values:** A configured ``MongoClient``
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:26]: Suggestion: separate these two lines to make it easier to scan (using the "|" rST formatter).

File: source/includes/auto-encryption-options.rst [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+   * - **tlsOptions**
+     -  **Type:** Object
+
+       **Description:**
+       Options to configure ``TLS`` connections to KMS providers.
+
+       To learn more about ``tlsOptions``, see the `API documentation <{+api+}/interfaces/AutoEncryptionTlsOptions.html>`__.
+
+       **Default:** ``undefined``
+       **Accepted Values:** An object containing the TLS options.
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:64]: Issue: indentation is off by a space so it's not rendering

File: source/includes/auto-encryption-options.rst [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+       **Default:** ``undefined``
+       **Accepted Values:** An object containing the encryption schema.
+
+   * - **bypassQueryAnalysis**
+     - **Type:** Boolean
+
+       **Description:**
+       Disables automatic analysis of outgoing commands. Set ``bypassQueryAnalysis``
+       to ``true`` to use explicit encryption on indexed fields without the
+       ``crypt_shared`` library. Defaults to ``false`` if not specified.
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:84]: Suggestion: define both branches of behavior, e.g.
When set to true, the driver uses explicit encryption ...
When set to false, the driver ...

Similar comment for cryptSharedLibRequired.

File: source/includes/auto-encryption-options.rst [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+   * - Name
+     - Description
+
+
+   * - **cryptSharedLibPath**
+     - **Type:** String
+
+
+       **Description:**
+       The path to the ``crypt_shared`` library.
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:114]: Issue: uncertain whether this is the path containing the library or the path of the library.
E.g. 
Path containing: /usr/local/bin
Path to: /usr/local/bin/crypt_shared

**Review by ccho-mongodb — COMMENTED** [review]
Will review the rest when I get back later tonight.

File: source/includes/auto-encryption-options.rst (hunk 10 above)
Comment (ccho-mongodb) [reply source/includes/auto-encryption-options.rst:26]: E.g. https://github.com/mongodb/docs-kafka-connector/blob/master/source/source-connector/configuration-properties/change-stream.txt#L39-L43

File: source/fundamentals/encrypt-fields/qe.txt [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      :start-after: begin client encryption
+      :end-before: end client encryption
+      :language: javascript
+      :dedent:
+
+Further Reading
+---------------
+
+For thorough coverage of the core concepts, use-cases, examples, and
+reference, visit the main {+qe+} :ref:`documentation site <qe-manual-feature-qe>`.
```
Comment (ccho-mongodb) [inline source/fundamentals/encrypt-fields/qe.txt:80]:
```suggestion
For coverage of the core concepts, use-cases, examples, and
reference, visit the :ref:`{+qe+} documentation site <qe-manual-feature-qe>`.
```

https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/#link-and-cross-reference-examples

File: source/fundamentals/encrypt-fields/qe.txt [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+encrypted data.
+
+
+Configuration
+-------------
+
+To use {+qe+}, you must:
+
+- Configure the driver with specific settings for encryption.
+- Configure and use the driver's encryption package, ``mongodb-client-encryption``.
```
Comment (ccho-mongodb) [inline source/fundamentals/encrypt-fields/qe.txt:39]: Issue: potential to miss the connection between these items and the subheaders. Currently references "driver"  and "encryption package" which don't have clear connections to the titles of the subheaders.

Suggestion: Explicitly mention which configuration subheader to which each bullet point is related.

File: source/includes/auto-encryption-options.rst [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+       To learn more about {+cmk-long+}s, see :ref:`qe-reference-keys-key-vaults`.
+
+       **Accepted Values:** An object containing the KMS provider or providers.
+
+   * - **tlsOptions**
+     -  **Type:** Object
+
+       **Description:**
+       Options to configure ``TLS`` connections to KMS providers.
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:59]: Issue: TLS shouldn't be monospace. Applies to all instances in which it's used to describe the mechanism rather than value/parameter/argument/etc.

File: source/includes/auto-encryption-options.rst [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+       To learn more about ``tlsOptions``, see the `API documentation <{+api+}/interfaces/AutoEncryptionTlsOptions.html>`__.
+
+       **Default:** ``undefined``
+       **Accepted Values:** An object containing the TLS options.
+
+   * - **encryptedFieldsMap**
+     - **Type:** Object
+
+       **Description:**
+       An encryption schema.
```
Comment (ccho-mongodb) [inline source/includes/auto-encryption-options.rst:70]: Suggestion: not necessary, but could be helpful to provide more detail on what the schema contains: which fields to encrypt, type information, and types of queries supported.

Perhaps unrelated, but the page to which it's linked doesn't describe it as a map but rather as encryptedFieldObject in the code.

Comment (terakilobyte) [reply source/includes/auto-encryption-options.rst:70]: Hmm. I've added more helpful text but I'm thinking about the second part of the comment.
